main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1483,31 +1483,44 @@ object Resolver {
           ResolvedAst.Expr.FixpointQueryWithProvenance(es, s, withh, loc)
       }
 
+    case NamedAst.Expr.FixpointQueryWithSelect(exps, queryExp, selects, from, where, pred, loc) =>
+      val esVal = traverse(exps)(resolveExp(_, scp0))
+      val qeVal = resolveExp(queryExp, scp0)
+      // We cannot call resolvePredicateBody as it does not allow new variables to be introduced
+      val fVal = traverse(from) {
+        case NamedAst.Predicate.Body.Atom(pred0, den, polarity, fixity, terms, loc1) =>
+          val ts = terms.map(resolvePattern(_, scp0, ns0, root))
+          Validation.Success(ResolvedAst.Predicate.Body.Atom(pred0, den, polarity, fixity, ts, loc1))
+        case _ => throw InternalCompilerException("Unexpected predicate body when expecting body atom", loc)
+      }
+      flatMapN(esVal, qeVal, fVal) {
+        case (es, qe, f) =>
+          val ts = f.flatMap(_.terms)
+          val scp1 = ts.foldLeft(scp0)(
+            (acc, t) => t match {
+              case ResolvedAst.Pattern.Var(sym, _) => acc ++ mkVarScp(sym)
+              case _ => acc
+            })
+          val sVal = traverse(selects)(resolveExp(_, scp1))
+          val wVal = traverse(where)(resolveExp(_, scp1))
+          mapN(sVal, wVal) {
+            case (s, w) =>
+              ResolvedAst.Expr.FixpointQueryWithSelect(es, qe, s, f, w, pred, loc)
+          }
+      }
```

Comment:
The above code here for this case is abit difficutl to follow. Maybe it can be restructured or some helper methods introduced?